### ISSUE TITLE
Revamp Virtual File System

### ISSUE THREAD

**Issue by Wildenhaus** [xenia-project/xenia#1293]
After working with VFS for a bit, its limitations have become increasingly apparent. Contributors have done a great job to lay a foundation for I/O operations, but in its current state, we are backed into a corner. I am under the impression that VFS could use a revamp, or possibly even a ground-up re-write for the following reasons:

- **It's too closely coupled with the host OS's filesystem**
Right now, we have `HostPathDevice` which allows us to read and write from the host filesystem. However, it seems like every other device has been designed to essentially do the same thing. They map a file from the host and mount it. At the most basic level, this works fine, but _it blocks us from being able to open or mount files from another device_. Many games have STFS content stored inside of the game's image, and unless we extract it to the host and then mount it, it's inaccessible.

- **It doesn't take into account writing data to a compressed filesystem**
Right now only `HostPathDevice` implements writing files, and that's handled inside of `file.h`. _If we want the functionality of being able to write to profiles or save games in the 'correct' format, we'll need to be able to handle it in a more robust manner_. For some virtual devices, we will need to consider adding sectors/blocks to a filesystem, and potentially even rehash and resign those files. This cannot be feasibly done from a file class in most cases.

- **It doesn't emulate how the Xbox handles its filesystem**
The stated goal of this project is to "completely reverse engineer" the Xbox 360. Thus, it would make sense to design VFS in a way that resembles how the Xbox might do it. What we have right now works just fine, but if we really want to build a robust project that lives up to its goal, we need to change the way in which we approach it.

- **tl;dr**
A lot of commonly requested features (DLC/TU/Profiles) are going to require functionality that isn't feasible with the current VFS implementation, and I would like to propose a re-design.

**Comment by eadmaster** [comment]
>  If we want the functionality of being able to write to profiles or save games in the 'correct' format, we'll need to be able to handle it in a more robust manner.

so, writing game saves in the native format is planned?
(the `current conversion process is quite complicated](https://www.reddit.com/r/xenia/comments/hg1yc2/how_to_convert_xbox360s_save_data_to_xenias_save/))
